### ISSUE TITLE
examples(rhai): Building surrogate cache keys

### ISSUE THREAD

**Issue by abernix** [apollographql/router#648]
Let's get an example using Rhai script that allows building a surrogate response cache key from subgraph responses.  e.g.:subgraph `a` returns `cache-key: one-way` and subgraph `b` returns `cache-key: or-another`, and the response key returned to the client is `surrogate-cache-key: one-way,or-another`

blocked by #661

Worth noting that CAN actually be done as a Rust plugin today, but we just want to make it more simple and allow it in Rhai.

**Comment by Geal** [comment]
right now we do not store response headers anywhere and have not thought about how and where they could be stored, so I propose we put off this example until we solve that

**Comment by Geal** [comment]
this should be doable now that rhai plugins have access to subgraph response headers

**Comment by kroehre** [comment]
Looking forward to this example, I'm not seeing subgraph response headers in any of the rhai plugin response objects

**Comment by garypen** [comment]
Most responses expose `response.headers`. This will contain the response headers which will eventually be propagated back to the client. The `Subgraph` `response.headers` can be read to examine headers returned by a subgraph.

I'm going to guess that the existence of `request.sub_headers` on the `SubgraphRequest` is the source of confusion...

In a `Request`, `headers` are the headers which were originally supplied by the client (possibly modified during the progression of a plugin pipeline). For `SubgraphRequest` we also need a way to indicate the headers that we are going to send to downstream subgraphs, which is what `request.sub_headers` represents on the `SubgraphRequest`. Is that helpful?

**Comment by kroehre** [comment]
Hey thanks for the reply! For some reason `response.headers` is null when I `print` them from the `subgraph_service` hook callback:
```javascript
fn subgraph_service(service, subgraph) {
  const response_callback = Fn("process_response");
  service.map_response(response_callback);
}

fn process_response(response) {
  print(`${response.headers}`);
}
```
In our case the goal is to propagate a few specific headers from subgraph responses.

**Comment by garypen** [comment]
Three options:
 - the subgraph isn't returning any headers
 - the headers aren't being propagated from the router subgraph handling
 - the rhai language binding has a bug which means it isn't picking up the headers properly.

I'll work through them and see if I can figure out what is happening.

**Comment by garypen** [comment]
And the winner is:

 - the headers aren't being propagated from the router subgraph handling

I've raised #1094 and will look into this.

**Comment by garypen** [comment]
@kroehre I've just merged the fix for #1094. This is now available on `main` and will be in the next release of the router.

**Comment by kroehre** [comment]
@garypen nice work! thank you
